tidy3d/plugins/invdes/function.py

```diff
@@ -0,0 +1,78 @@
+# post processing function utilities
+
+import typing
+
+import tidy3d as td
+import tidy3d.plugins.adjoint as tda
+import jax.numpy as jnp
+
+
+class Utilities:
```

Comment:
Is there a reason why "functional" utilities are contained within a class? As there is no state, having these as functions in a module might be cleaner perhaps?

Reply:
Yea so originally I had these as individual functions, but it becomes a bit annoying to manage:
* The documentation needs to add all of the functions one by one, they aren't really organized.
* They aren't inside of a contained namespace (like `invdes.Utils`) so it kind of clutters the `invdes` namespace.

It seemed nicer to just contain them inside of a class, but I'm not sure, maybe we can just use the module itself as the namespace and just do `from invdes.functions import _`.

Do you have any strong opinions either way? I'm not sold on a particular approach, but do think grouping them in a namespace is generally nicer than free floating functions

Reply:
I don't really know what's better between doing it through class methods or just through a module. Note a similar discussing #1616 @dmarek-flex where I recommended class methods. Maybe we can think / discuss pros and cons a bit here

Reply:
I can't really comment regarding the documentation since I'm not too familiar with how that is handled here.
From a purely prgramming-related perspective though, a class (to me) is something that bundles code and state/data, where ideally the code doesn't "make sense" without the data. So the purist in me would say that even `staticmethod`s are a code smell (not to say they aren't useful sometimes).
If it's just about namespacing, then I think modules work fine? That depends mostly on how it's handled in `__init__.py`. I see that currently the whole module is imported into the `invdes` namespace via `from .function import Utilities`. If this is changed to just `from . import function`, then the functions in that module would be callable (after importing `invdes`) as, e.g., `function.make_array()`, so there would still be a separation of namespaces if so desired. Should rename `functions.py` to `utilities.py` then, probably.
Not a strong opinion, but an opinion 😅

Reply:
In my case, I have a lot of supporting functions that users should not need to use, so I would like to hide them somewhat. I guess that could be accomplished by modifying my `__init__.py` to be more selective. In C++, I would have probably used a namespace instead of class as well, so maybe that is a clue!

Reply:
If we went with module, we could just keep the private functions with a leading underscore and not imported in `__init__.py`. I also am starting to come around to this idea. my main concern is just wondering if it will be straightforward for the user to explore which functions are available in the module. I guess if, eg. `import invdes.utils as utils`, they could call `dir(utils)`, which doesn't look good, but apparently `help(utils)` might be formatted quite well for us. Perhaps this is fine. Maybe I'll give it a try.

Reply:
Went with a module approach. Seems fine to me, will keep it. Thanks for the suggestions @yaugenst.